operator/pkg/service/envoy_config.go

```diff
@@ -43,14 +56,271 @@ func newEnvoyConfigManager(maxRetries int) (*envoyConfigManager, error) {
 }
 
 // getByKey is a wrapper of Store.GetByKey but with concrete Endpoint object
-func (em *envoyConfigManager) getByKey(key string) (*v2.CiliumEnvoyConfig, bool, error) {
+func (em *envoyConfigManager) getByKey(key string) (*ciliumv2.CiliumEnvoyConfig, bool, error) {
 	objFromCache, exists, err := em.store.GetByKey(key)
 	if objFromCache == nil || !exists || err != nil {
 		return nil, exists, err
 	}
-	envoyConfig, ok := objFromCache.(*v2.CiliumEnvoyConfig)
+	envoyConfig, ok := objFromCache.(*ciliumv2.CiliumEnvoyConfig)
 	if !ok {
 		return nil, exists, fmt.Errorf("got invalid object from cache")
 	}
 	return envoyConfig, exists, err
 }
+
+func getEnvoyConfigForService(svc *slim_corev1.Service) (*ciliumv2.CiliumEnvoyConfig, error) {
+	resources, err := getResources(svc)
+	if err != nil {
+		return nil, err
+	}
+	return &ciliumv2.CiliumEnvoyConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       ciliumv2.CECKindDefinition,
+			APIVersion: ciliumv2.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", ciliumEnvoyLBPrefix, svc.GetName()),
+			Namespace: svc.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: slim_corev1.SchemeGroupVersion.String(),
+					Kind:       "Service",
+					Name:       svc.GetName(),
+					UID:        svc.GetUID(),
+				},
+			},
+		},
+		Spec: ciliumv2.CiliumEnvoyConfigSpec{
+			Services: []*ciliumv2.ServiceListener{
+				{
+					Name:      svc.GetName(),
+					Namespace: svc.GetNamespace(),
+				},
+			},
+			Resources: resources,
+		},
+	}, nil
+}
+
+func getResources(svc *slim_corev1.Service) ([]ciliumv2.XDSResource, error) {
+	var resources []ciliumv2.XDSResource
+	listener, err := getListenerResource(svc)
+	if err != nil {
+		return nil, err
+	}
+	resources = append(resources, listener)
+
+	routeConfig, err := getRouteConfigurationResource(svc)
+	if err != nil {
+		return nil, err
+	}
+	resources = append(resources, routeConfig)
+
+	clusters, err := getClusterResources(svc)
+	if err != nil {
+		return nil, err
+	}
+	resources = append(resources, clusters...)
+	return resources, nil
+}
+
+func getClusterResources(svc *slim_corev1.Service) ([]ciliumv2.XDSResource, error) {
+	cluster := &envoy_config_cluster_v3.Cluster{
```

Comment:
Question: Is there eventually going to be a way to configure lower level Envoy configs via the Service resource?

Reply:
I think this PR is kind of a first step to achieve the above i.e. configuring envoy resource from k8s service annotation. 

I just add 2 simple mutation methods based on service annotation, but we can easily add more.